File: gitlfs-client/src/test/java/ru/bozaro/gitlfs/client/HttpReplay.java
package ru.bozaro.gitlfs.client;

import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpUriRequest;
import org.testng.Assert;

import javax.annotation.Nonnull;
import java.io.IOException;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

/**
 * Replay recorded HTTP requests.
 *
 * @author Artem V. Navrotskiy
 */
public class HttpReplay implements HttpExecutor {
  @Nonnull
  private final Deque<HttpRecord> records;

  public HttpReplay(@Nonnull List<HttpRecord> records) {
    this.records = new ArrayDeque<>(records);
  }

  @Nonnull
  @Override
  public CloseableHttpResponse executeMethod(@Nonnull HttpUriRequest request) throws IOException {
    final HttpRecord record = records.pollFirst();
    Assert.assertNotNull(record);

    final String expected = record.getRequest().toString();
    final String actual = new HttpRecord.Request(request).toString();
    Assert.assertEquals(actual, expected);

    return record.getResponse().toHttpResponse();
  }

  public void close() {
    Assert.assertTrue(records.isEmpty());
  }
}

File: gitlfs-client/src/test/kotlin/ru/bozaro/gitlfs/client/HttpReplay.kt
package ru.bozaro.gitlfs.client

import org.apache.http.client.methods.CloseableHttpResponse
import org.apache.http.client.methods.HttpUriRequest
import org.testng.Assert
import java.io.IOException
import java.util.*

/**
 * Replay recorded HTTP requests.
 *
 * @author Artem V. Navrotskiy
 */
class HttpReplay(records: List<HttpRecord>) : HttpExecutor {
    private val records: Deque<HttpRecord>

    @Throws(IOException::class)
    override fun executeMethod(request: HttpUriRequest): CloseableHttpResponse {
        val record = records.pollFirst()
        Assert.assertNotNull(record)
        val expected = record.request.toString()
        val actual = HttpRecord.Request(request).toString()
        Assert.assertEquals(actual, expected)
        return record.response.toHttpResponse()
    }

    override fun close() {
        Assert.assertTrue(records.isEmpty())
    }

    init {
        this.records = ArrayDeque(records)
    }
}
